### ISSUE TITLE
[Enhancement]: Analytics Revenue should consider order paid date instead of order created date.

### ISSUE THREAD

**Issue by AashikP** [woothemes/woocommerce#32693]
### Revenue

* In Analytics and Reports, the payment revenue is assigned to the date when an order is created rather than the date when a payment was made.
* If an order is created on January 1st but it is only paid for on February 1st - I'd expect the reports to show the revenue on February 1st.
* Whenever the payment is made, reports will show it as if it happened on the same day as when the order was created. 

### Describe the solution you'd like

Perhaps add an option to let the merchant choose if they want the reports assigned to the order date, or the actual payment date?

### Additional context

* Originally requested here: https://github.com/woocommerce/woocommerce-admin/issues/6535 cc @grig-23 
* Slack: p1650010691719899-slack-C7U3Y3VMY
* Reports: 4949047-zen

**Comment by AashikP** [comment]
This is also helpful for merchants where manual orders are created well in advance, and their customers pay a couple of weeks later. As requested in 5461323-zen:

> When we send a manual order clients sometimes wait to pay meaning it might be 2 weeks or a month which makes the sales an actual sale within the next month, but instead of updating the order date to be payment date Woo allows the date to stay the same which means that the order is counting for the previous month and NOT the month in which the payment is made

**Comment by imodouglas** [comment]
The merchant came back stating that this is important for WC users that rely on invoices for tax purposes and can lead to an intentional tax fraud case. → 5461323-zen

Can this be looked into any sooner?

Merchant's reply:
> I honestly think this is of much more importance than a simple request for development in the future. You realize that the system is back dating transactions which is intentionally causing transactions to be back dated versus when they were paid which is giving inaccurate information for taxing purposes. This would require every Ecommerce company that uses WooCommerce invoicing on the back end to have to not only file taxes monthly/quarterly but then have to go back and amend taxes for previous months so that it accurately reflects the correct totals. Having these totals not be filed in the correct month and or bouncing back to previous months while not knowing and if audited can be considered committing tax fraud and should be taken to the highest importance and dealt with urgently.

**Comment by AnnaMag** [comment]
@joelclimbsthings could we start with estimating the effort, so we can prioritize it correctly?

**Comment by joelclimbsthings** [comment]
@AnnaMag I'd say the estimate will depend on how we want to handle this. Do we indeed want to make this an option, or could the case be made that we should just change this across the board? Is there a use case to preserve how it's calculated currently? (I'm not familiar with the implications here). 

It looks like the original issue was pinging @LevinMedia to bottom out on this without reply. 

cc @pmcpinto in case you have an opinion or know who to contact?

**Comment by PowerGPU** [comment]
@joelclimbsthings @AnnaMag @pmcpinto I am the client who brought up the concern and hope that the team can truly set this as a priority. I tried to find a good reason why any company using the platform would prefer to have the estimate date rather than the payment date and I honestly can't find any. Ideally if it's easier for the development team to differentiate the estimate date and the payment date by placing the information in two different areas then that can work but then all reporting would need to be tied to that payment date. For tax purposes the transactions need to reflect the payment date.  
    We are a computer manufacturing company who sells custom PCs online. We have noticed two problems with the date stamps when doing manual orders. 1) When we send an estimate to a client it could be a few weeks to months until they decide to move forward with their order and pay. During this time the invoices goes through multiple change requests by the client before final purchase. The issue is that when the client pays their order it leaves the clients order as the oldest order since the date does not update to the payment date. This in turn makes the client skip the line since we fulfill orders on oldest to newest. We have a slew of orders that paid for example in the month of August. If the client pays their estimate for an estimate they received in July then  they immediately jump the line as the oldest order and get their order first which is not correct.
2) When filing sales and use taxes for each state, particular states have us file on a monthly basis due to the amount of the amount of revenue made in that state per month. I noticed this big problem when I went back to verify I filed the previous month correctly only find that there were more orders that were not accounted for. This immediately horrified me because now if I ran the reports needed to file with the state then it looks like I intentionally omitted the transaction to avoid paying sales and use tax. When I researched the transaction apparently this client was given an estimate and didn't pay it until a month or so later so the recent payment posted in the past as a transaction and so the reports showed different amounts. This immediately was a red flag since this would make us look like we are committing insurance fraud God forbid someone audits this. If I were to manually change this then this would also results having to go back to previous months, check totals versus new totals, then find the transaction and change the date, then hope that the new information gets updated in Avalara correctly to show the correct date, which I doubt since the transaction was already sent to Avalara. At the end of the day it's causing a huge financial and taxation problem. I am honestly surprised that this was set up this way. I ask the team to please have this set up as a priority to avoid all WooCommerce customers from  have continued tax issues.

**Comment by pmcpinto** [comment]
Hi @PowerGPU thanks for your detailed feedback. By payment date, do you mean the order completion date?

**Comment by pmcpinto** [comment]
> @AnnaMag I'd say the estimate will depend on how we want to handle this. Do we indeed want to make this an option, or could the case be made that we should just change this across the board? Is there a use case to preserve how it's calculated currently? (I'm not familiar with the implications here).

@joelclimbsthings is it possible to have a high-level estimate of the two approaches?

**Comment by PowerGPU** [comment]
@pmcpinto No I mean the payment received date. Because we are a manufacturing company our completion date can sometimes be weeks or months later. What I meant with the dates above in the description was the payment date, when the client paid for their order.

**Comment by pmcpinto** [comment]
thanks for the clarification. I've asked this because as far as I know in some use cases it's not possible to record the payment received date. For example, if the payment is made through offline payment options such as cash on delivery. cc @joelclimbsthings

**Comment by joelclimbsthings** [comment]
cc @octaedro since he's the current porter.

**Comment by PowerGPU** [comment]
@pmcpinto That is a valid point which has presented a problem with the QuickBooks integration. Currently Woo has no possible way of posting a Wire or ACH payment and because of this these transaction do not come into QuickBooks. Realistically why is it that Woo commerce does not post the paid date as the order date? An order becomes an order once it's paid correct?

**Comment by octaedro** [comment]
I'm not sure if I understood the problem correctly.
I'll estimate the options I see:

#### Currently we are saving the completion date as the payment date.
The changes to use that information instead of the order's creation date in the report are pretty small, and they could be done in one day, maybe two.

#### If we would want to use another order state (E.g., the date when the order was changed to `Processing`, shown by default in the report) to use it in the `Analytics > Revenue` report, we should:
- Add a column `date_processed` to the table `wp_wc_order_product_lookup` (the option above will need something similar too).
- Add a method to save the information in the DB to the hook `woocommerce_order_status_processing` (or add the code that we need to the method [status_transition](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-order.php#L356)).
- Show the `date_processed` instead the `date_created` in `Analytics > Revenue`.

I think that adding those changes, plus tests, etc would take 2 or maybe 3 days.

#### Adding an input to save the estimated payment date to the order.
It would look similar to this

![screenshot-test local-2022 09 01-16_09_03](https://user-images.githubusercontent.com/1314156/187993601-4c91cd74-30c6-4596-b5f0-6f616e6eb24f.png)

I think that adding the changes needed here (pretty similar to what we would need in the last option) probably would take 3 days.

// cc @pmcpinto @AnnaMag

**Comment by PowerGPU** [comment]
@pmcpinto Why wouldn't the payment date be recorded is my point. The payment date is technically the order date, the date the client chose to purchase and is waiting for their order to be delivered. Shopify counts this date as the order date, once payment is received. Why would it be done otherwise? No matter what way the client pays the payment date should be captured as the order date.

**Comment by PowerGPU** [comment]
@octaedro You mentioned you are currently saving the completion date as the payment date? Why? The completion date of the order, meaning the order has been shipped to the client is when the order is completed. The date the customer pays is the order commence date, when the client paid for the order and is awaiting to receive the item. Having this correct all issues with the current set up. 
Having the payment date save as the order date would solve all the taxation issues, in any Ecommerce scenario whether it's a single item purchase then shipment or as in our case a purchase date then the manufacturing process starts and the order is fulfilled between a few weeks to maybe over a month later then the information is recorded correctly. 

Here is an example of the scenario I mean: 
1) Client requested a computer. Our team sent them an estimate on 08/29/2022. The client chose to not pay the order until 09/02/2022. The client technically paid and processed their order on 09/02/2022. When the order comes over to Avalara the order is coming over as the date that the invoice was created instead of the payment date which is the actual client purchase date. 
Woo
![Capture](https://user-images.githubusercontent.com/112298784/188779808-efc7d3c5-9e14-4124-b846-99003c16dc31.JPG)
Avalara
![Capture 2](https://user-images.githubusercontent.com/112298784/188779957-4075206d-e005-4410-937f-92be9976461e.JPG)

Now because the order was paid in September and taxes were collected in September it should show up as a September order but it does not. Now when I run my reports to file my sales and use taxes I would be committing tax fraud filing like this meaning I cannot trust Woo nor Avalara reporting at this point. This transaction should NOT show up for the month of August but it does and that is because the system is grabbing the invoice creation date as the order date and the payment date which is incorrect.  @pmcpinto @octaedro 

In other cases we have seen the invoices back date to almost two months back meaning now the taxes I filed those months are incorrect and if I were ever tax audited it would be a problem even though the fault was not mine. I like accuracy and prefer not to get fined by the government lol. Thanks for the help everyone and hope the changes could be done as soon as possible.

**Comment by PowerGPU** [comment]
At the end of the day the order date which is the payment date should not be something that can be edited either way. When a payment posts to the account regardless of any other scenario it shouldn't change. Having this as a customizable option can also cause internal issues for any business. Example: An employee of mine has a friend that wants his order ASAP but technically should be in line with everyone else. The employee can easily change the order date and then have the friends order expedited without any consequences to actions. This leave too much room for error. Can anyone think of a reason why a payment posting to the account and having an official ORDER DATE posted without being able to be edited because at the end of the day the date is something permanent that shouldn't be changed but instead sealed once the client pays their order. 

This segways to my other concern that the system allows the status of paid (processing) to be edited. This is a huge concern. Coming previously from Shopify, when a client paid the order was set to paid and it's not something you can edit unless you wanted to cancel the order or refund. Being able to change the status of an order also leads to fraud issues. 
Example: One of my employees decides that he can't afford a computer and wants one. He easily puts in a friends information so it's not suspected as his computer. He puts in the order and then simply flips the switch of pending payment to processing. There is no indications that the order is ACTUALLY paid or alerts that go off on the order. The order can go through its entirety in the build process and shipping without anyone knowing any better. The picker just picks the parts, the builder and software team just have printed copies of the invoice which doesn't indicate paid status and finally shipping just prints labels not confirming payment. In the blink of an eye a PC is shipped out without any payment and the employee has stolen a PC. 

I know these examples involve a lot of employee theft and God forbid this happens since I have an amazing crew but these are legitimate concerns seeing that the company grows year over year and not at small steps. I guess the question is why would paid status (processing) be something that someone can edit? As mentioned in Shopify and other platforms there is no way to change that.

**Comment by islevista** [comment]
Hi there, I have been following this post and I absolutely agree that this needs to be updated. 
This is crucial for my business as well. Please let us know once it is fixed, thank you.

**Comment by pmcpinto** [comment]
> @pmcpinto Why wouldn't the payment date be recorded is my point. The payment date is technically the order date, the date the client chose to purchase and is waiting for their order to be delivered. Shopify counts this date as the order date, once payment is received. Why would it be done otherwise? No matter what way the client pays the payment date should be captured as the order date.

The order date should be the payment date if we could accurately track all the payments, but I'm not sure it will be possible due to methods like COD. And we also need to consider businesses when the payment doesn't happen when the order is placed. Therefore, I think the best approach is to allow the user to choose the order status that should be considered in the report: when the order is created, when is marked as completed, etc @octaedro @jarekmorawski let me know what you think

**Comment by PowerGPU** [comment]
@pmcpinto If the case is COD then let this status be sealed as a payment method and continue to make that COD date as the order date. COD is a method of payment at the end of the day the same as a VISA or PayPal option.

**Comment by islevista** [comment]
I am currently in a situation where this is completely throwing off my invoices with our manufacturing office. 

They use our WooCommerce Analytics Reports to look at the orders received and charge us based on the orders on the report from the previous week. This is usually fine, however if we have one of the aforementioned situations, it causes a discrepancy on the invoices. 

For example:
-A customer placed an order in May but the payment failed. 
-We attempted to reach out to them but it was not until September that the customer went onto our website to update his payment method and have the order be approved and switch from "Failed" to "Processing'. 
-When our manufacturer looks at the Analytics report for that week in September however, this order will not be there because it is still under its creation date, from May. 
-Thus, the order goes by undetected and we are never charged for it. 

We don't know how many of these orders have slipped by over the years and it's extremely frustrating. 

+All I am asking is for an option to have my orders be viewed not from the "Date Created" but from when we actually received the payment. Thank you.

**Comment by PowerGPU** [comment]
@pmcpinto @AnnaMag Just wanted to follow up. Was reassured by support that this would be prioritized do to it falsely reporting taxes for wrong dates due to not accurately applying the paid date as the sales order date. Seems like the conversation has grown cold.

**Comment by AnnaMag** [comment]
@PowerGPU thank you for following up on this. We will get back to you on this this week.

**Comment by AnnaMag** [comment]
@PowerGPU thank you for following up on this. 

We have not yet settled on a solution that would be encompass all business cases. We understand the importance of the issue and will get back to you all on this asap.

**Comment by pmcpinto** [comment]
Hi everyone, thanks for the feedback shared so far. We will change the date used in the revenue report to when the order status changed to processing, which [happens when the payment is successful](https://woocommerce.com/document/managing-orders/#understanding-order-statuses-in-relation-to-payment-gateways). We'll also introduce an option to change the order status used in the report.
We've explored other options, but they could potentially impact extensions that leverage the order data, so we'll not make any changes in this area for the time being.
We'll open another issue next week with the implementation details.

**Comment by PowerGPU** [comment]
@pmcpinto I would hope that it follows the payment and not necessarily the change in payment status to processing. I say this because sometimes orders are placed on hold then placed back to processing which would then keep moving these dates which can cause a problem. The overall order date needs to tie to the payment itself and not the status because you also have apps that allow status changes which this can potentially cause an issue?

**Comment by PowerGPU** [comment]
@pmcpinto Has the team also been able to discuss the details on the payment status changes overall? As mentioned before there should be no reason why a payment status should be editable unless you are cancelling/ refunding a client. The ability to change a payment status' leaves open the ability for staff/employees or anyone internally who has access to the system to commit fraud. There is currently no system in place that Woo has that prompts that an order has or has no payment attached. So in essence an employee can mark an order as paid even if a payment is not truly attached then just have the products shipped without anyone noticing. This is a huge liability and unfortunately there is no way to even run a report to catch if any orders have shipped with no payment attached. As mentioned before there is no reason why someone order payment status should change and is something that should be locked unless the order is cancelled and a return is processed.

**Comment by pmcpinto** [comment]
> @pmcpinto I would hope that it follows the payment and not necessarily the change in payment status to processing. I say this because sometimes orders are placed on hold then placed back to processing which would then keep moving these dates which can cause a problem. The overall order date needs to tie to the payment itself and not the status because you also have apps that allow status changes which this can potentially cause an issue?

Correct. We revisited the solution, and we'll use the `date_paid`. I'll close this issue, as we created a new one with the implementation details: #35348

> @pmcpinto Has the team also been able to discuss the details on the payment status changes overall? As mentioned before there should be no reason why a payment status should be editable unless you are cancelling/ refunding a client. The ability to change a payment status' leaves open the ability for staff/employees or anyone internally who has access to the system to commit fraud. There is currently no system in place that Woo has that prompts that an order has or has no payment attached. So in essence an employee can mark an order as paid even if a payment is not truly attached then just have the products shipped without anyone noticing. This is a huge liability and unfortunately there is no way to even run a report to catch if any orders have shipped with no payment attached. As mentioned before there is no reason why someone order payment status should change and is something that should be locked unless the order is cancelled and a return is processed.

Thanks for flagging this situation. This will require further exploration, but we're not sure when it will be possible to do it. We'll open a separate issue focused on this use case.
